File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/VerifiableJsonLdBuilder.java
package city.smartb.iris.ldproof;

import city.smartb.iris.jsonld.JsonLdObject;
import city.smartb.iris.ldproof.util.CanonicalizationUtil;

import java.util.LinkedHashMap;
import java.util.Map;

public class VerifiableJsonLdBuilder {

    public static VerifiableJsonLdBuilder builder(Map<String, Object> json) {
        return new VerifiableJsonLdBuilder(json);
    }

    private final LinkedHashMap<String, Object> json;

    public VerifiableJsonLdBuilder(Map<String, Object> json) {
        this.json = new LinkedHashMap<>(json);
    }

    public String buildCanonicalizedDocument() {
        removeProofFromJsonLdObject(json);
        return CanonicalizationUtil.buildCanonicalizedDocument(json);
    }

    public static void removeProofFromJsonLdObject(LinkedHashMap<String, Object> json) {
        json.remove(LdProof.JSON_LD_PROOF);
    }

    public VerifiableJsonLd build(LdProof proof) {
        this.json.put(LdProof.JSON_LD_PROOF, proof.asJson());
        return  new VerifiableJsonLd(this.json);
    }
}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/VerifiableJsonLdBuilder.kt
package city.smartb.iris.ld.ldproof

import city.smartb.iris.ld.ldproof.util.CanonicalizationUtil

class VerifiableJsonLdBuilder(json: Map<String, Any>) {
    private val json: LinkedHashMap<String, Any>

    init {
        this.json = LinkedHashMap(json)
    }

    fun buildCanonicalizedDocument(): String {
        removeProofFromJsonLdObject(json)
        return CanonicalizationUtil.buildCanonicalizedDocument(json)
    }

    fun build(proof: LdProof): VerifiableJsonLd {
        json[LdProof.JSON_LD_PROOF] = proof.asJson()
        return VerifiableJsonLd(json)
    }

    companion object {
        fun builder(json: Map<String, Any>): VerifiableJsonLdBuilder {
            return VerifiableJsonLdBuilder(json)
        }

        fun removeProofFromJsonLdObject(json: LinkedHashMap<String, Any>) {
            json.remove(LdProof.JSON_LD_PROOF)
        }
    }
}
